lib/spack/spack/spec.py

```diff
@@ -1829,6 +1805,43 @@ def concretize(self):
             if s.namespace is None:
                 s.namespace = spack.repo.repo_for_pkg(s.name).namespace
 
+            # Add any patches from the package to the spec.
+            patches = []
+            for cond, patch_list in s.package_class.patches.items():
+                if s.satisfies(cond):
+                    for patch in patch_list:
+                        patches.append(patch.sha256)
+            if patches:
+                # Special-case: keeps variant values unique but ordered.
+                s.variants['patches'] = MultiValuedVariant('patches', ())
+                mvar = s.variants['patches']
+                mvar._value = mvar._original_value = tuple(dedupe(patches))
+
+        # Apply patches required on dependencies by depends_on(..., patch=...)
+        for dspec in self.traverse_edges(deptype=all,
+                                         cover='edges', root=False):
+            pkg_deps = dspec.parent.package_class.dependencies
+            if dspec.spec.name not in pkg_deps:
+                continue
+
+            patches = []
+            for cond, dependency in pkg_deps[dspec.spec.name].items():
+                if dspec.parent.satisfies(cond):
+                    for pcond, patch_list in dependency.patches.items():
+                        if dspec.spec.satisfies(pcond):
+                            for patch in patch_list:
+                                patches.append(patch.sha256)
+            if patches:
+                # note that we use a special multi-valued variant and
+                # keep the patches ordered.
+                if 'patches' not in dspec.spec.variants:
+                    mvar = MultiValuedVariant('patches', ())
+                    dspec.spec.variants['patches'] = mvar
+                else:
+                    mvar = dspec.spec.variants['patches']
+                mvar._value = mvar._original_value = tuple(
+                    dedupe(list(mvar._value) + patches))
```

Comment:
@tgamblin @scheibelp I'll write this comment here just to be sure I passed the information. What we are doing in the line:
```python
mvar._value = mvar._original_value = tuple(dedupe(list(mvar._value) + patches))
```
is breaking an invariant of the class (the assumption that `self._value` is totally ordered). This assumption is transparent to the user of the class, and bypassing the property setter here is not very robust. In fact any call to a method that does:
```python
self.value = self.value + other
```
can potentially reorder `self._value` and lead to #5587 . An example of such a method may be for instance `MultiValuedVariant.constrain`.